test/integration/provider/sessions_controller_test.rb

```diff
@@ -37,4 +37,21 @@ class Provider::SessionsControllerTest < ActionDispatch::IntegrationTest
 
     assert_redirected_to "http://example.net/?provider_id=#{account.id}&user_id=#{account.admin_user.id}"
   end
+
+  test "redirect users to SSO when there's only one authentication provider and enforce_sso is on" do
+  @provider.settings.update_column(:enforce_sso, true)
+  get new_provider_sessions_path
+  request = ActionDispatch::TestRequest.create
+  assert_redirected_to authorization_provider_bounce_path(authentication_provider.system_name)
+  end
+
+  test "does not redirect when there's more than one authentication provider" do
+  @provider.settings.update_column(:enforce_sso, true)
+  FactoryBot.create(:self_authentication_provider, account: @provider, published: true)
+  get new_provider_sessions_path
+  assert_response :success
+
+  # password login is disabled, there's no passoword login form
+  refute_match 'Email or Username', response.body
```

Comment:
This is very error-prone. `refute_match` is something to be avoided as much as possible, because it might not match for the wrong reasons, for example by just changing in the UI `Email or Username` for `Username or Email`.
If there is another way to check this, it is better 😄 Otherwise, have a test for when it must match and when doesn't, so if there is a change like the example I gave you, then at least we will detect the failure sooner.

Reply:
test is already there for match case in test/integration/sso_enforce_flow_test.rb
    # password signup form is visible
    get new_provider_sessions_path
    assert_response :success
    assert_match 'Email or Username', response.body